src/compact.jl

```diff
@@ -79,11 +81,21 @@ println(model)  # "Linear(3 => 1)"
 This can be useful when using `@compact` to hierarchically construct
 complex models to be used inside a `Chain`.
 """
-macro compact(fex, kwexs...)
-  # check input
+macro compact(fex, _kwexs...)
+  # check inputs
   Meta.isexpr(fex, :(->)) || error("expects a do block")
-  isempty(kwexs) && error("expects keyword arguments")
-  all(ex -> Meta.isexpr(ex, (:kw,:(=))), kwexs) || error("expects only keyword arguments")
+  isempty(_kwexs) && error("expects keyword arguments")
+  all(ex -> Meta.isexpr(ex, (:kw,:(=),:parameters)), _kwexs) || error("expects only keyword arguments")
+
+  # process keyword arguments
+  if Meta.isexpr(_kwexs[1], :parameters) # handle keyword arguments provided after semicolon
+    kwexs1 = map(ex -> ex isa Symbol ? Expr(:kw, ex, ex) : ex, _kwexs[1].args) 
+    _kwexs = _kwexs[2:end]
+  else
+    kwexs1 = ()
+  end
+  kwexs2 = map(ex -> Expr(:kw, ex.args...), _kwexs) # handle keyword arguments provided before semicolon
+  kwexs = (kwexs1..., kwexs2...)
```

Comment:
This logic handles the first case below, but can it be done nicely on all arguments to handle also the second case here?
```julia
julia> macro pr(exs...)
         @show exs; nothing
       end
@pr (macro with 1 method)

julia> @pr(; a=1, b) do x
         x+a+b
       end
exs = (:((x,)->begin
          #= REPL[57]:2 =#
          x + a + b
      end), :($(Expr(:parameters, :($(Expr(:kw, :a, 1))), :b))))

julia> @pr(x -> x+a+b; a=1, b)
exs = (:($(Expr(:parameters, :($(Expr(:kw, :a, 1))), :b))), :(x->begin
          #= REPL[58]:1 =#
          x + a + b
      end))

julia> @pr(x -> x+a+b, a=1, b=2)
exs = (:(x->begin
          #= REPL[59]:1 =#
          x + a + b
      end), :(a = 1), :(b = 2))
``` 
Or might it be better to have the function call itself with arguments in re-arranged order?

Reply:
Done. I added https://github.com/FluxML/Fluxperimental.jl/blob/9ab4b87f4ebbe646560675e9f65d8af28f082f84/src/compact.jl#L125 just so that the test could run, it should not stay very long hopefully:)

Reply:
Oops. That warning isn't sufficient because this PR doesn't have the new logic in #17. So I addressed this comment there instead (sorry for the mess...)

Reply:
Ok. So we merge this, and work on any further refinements in #17?

Reply:
Yes, that sounds good.